hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/shortcircuit/TestShortCircuitCache.java

```diff
@@ -909,4 +910,90 @@ public void testRequestFileDescriptorsWhenULimit() throws Exception {
       }
     }
   }
+
+  @Test
+  public void testDomainSocketClosedByDN() throws Exception {
+    BlockReaderTestUtil.enableShortCircuitShmTracing();
+    TemporarySocketDirectory sockDir = new TemporarySocketDirectory();
+    Configuration conf =
+        createShortCircuitConf("testDomainSocketClosedByDN", sockDir);
+    MiniDFSCluster cluster =
+        new MiniDFSCluster.Builder(conf).numDataNodes(1).build();
+    cluster.waitActive();
+    DistributedFileSystem fs = cluster.getFileSystem();
+    final ShortCircuitCache cache =
+        fs.getClient().getClientContext().getShortCircuitCache();
+    DomainPeer peer = getDomainPeerToDn(conf);
+    MutableBoolean usedPeer = new MutableBoolean(false);
+    ExtendedBlockId blockId = new ExtendedBlockId(123, "xyz");
+    final DatanodeInfo datanode = new DatanodeInfo.DatanodeInfoBuilder()
+        .setNodeID(cluster.getDataNodes().get(0).getDatanodeId()).build();
+    // Allocating the first shm slot requires using up a peer.
+    Slot slot1 = cache.allocShmSlot(datanode, peer, usedPeer, blockId,
+        "testReleaseSlotReuseDomainSocket_client");
+
+    cluster.getDataNodes().get(0).getShortCircuitRegistry()
+        .registerSlot(blockId, slot1.getSlotId(), false);
+
+    Slot slot2 = cache.allocShmSlot(datanode, peer, usedPeer, blockId,
+        "testReleaseSlotReuseDomainSocket_client");
+
+    cluster.getDataNodes().get(0).getShortCircuitRegistry()
+        .registerSlot(blockId, slot2.getSlotId(), false);
+
+    cache.scheduleSlotReleaser(slot1);
+
+    // make the DataXceiver timedout
+    Thread.sleep(5000);
+    cache.scheduleSlotReleaser(slot2);
+    Thread.sleep(10000);
```

Comment:
Can we make this sleep time shorter? Waiting for 15 seconds seems too excessive. You may have to change the timeout configuration.

Reply:
done